File: src/splitter_lrs.py
# This code is distributed under the GNU General Public License (see
# THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
# Copyright (C) 2014, the P2PSP team.

# {{{ Imports
from __future__ import print_function
import threading
import sys
import socket
import struct
import time
from color import Color
import common
from _print_ import _print_
from splitter_ims import Splitter_IMS
from splitter_dbs import Splitter_DBS
from splitter_fns import Splitter_FNS
from splitter_acs import Splitter_ACS
# }}}

# Lost chunk Recovery Set of rules
class Splitter_LRS(Splitter_ACS):
    # {{{

    def __init__(self):
        # {{{

        Splitter_ACS.__init__(self)

        # A circular array of messages (chunk_number, chunk) in network endian
        self.buffer = [""]*self.BUFFER_SIZE

        # }}}

    def print_modulename(self):
        # {{{

        sys.stdout.write(Color.yellow)
        print("Using LRS")
        sys.stdout.write(Color.none)

        # }}}

    def process_lost_chunk(self, lost_chunk_number, sender):
        # {{{

        Splitter_ACS.process_lost_chunk(self, lost_chunk_number, sender)
        message = self.buffer[lost_chunk_number % self.BUFFER_SIZE]
        peer = self.peer_list[0]
        self.team_socket.sendto(message, peer)
        #self.number_of_sent_chunks_per_peer[peer] += 1
        #self.destination_of_chunk[self.chunk_number % self.BUFFER_SIZE] = peer
        #number, chunk = struct.unpack(self.chunk_format_string, message)
        #chunk_number = socket.ntohs(number)
        sys.stdout.write(Color.cyan)
        print ("Re-sending", lost_chunk_number, "to", peer)
        sys.stdout.write(Color.none)

        # }}}

    def send_chunk(self, message, peer):
        # {{{

        Splitter_ACS.send_chunk(self, message, peer)
        self.buffer[self.chunk_number % self.BUFFER_SIZE] = message
        
        # }}}


    # }}}

File: src/core/splitter_lrs.cc
//
//  splitter_lrs.cc
//  P2PSP
//
//  This code is distributed under the GNU General Public License (see
//  THE_GNU_GENERAL_PUBLIC_LICENSE.txt for extending this information).
//  Copyright (C) 2016, the P2PSP team.
//  http://www.p2psp.org
//
//  LRS: Lost chunk Recovery Set of rules
//

#include "splitter_lrs.h"
#include "../util/trace.h"

namespace p2psp {
  using namespace std;
  using namespace boost;

  Splitter_LRS::Splitter_LRS() : Splitter_ACS(), buffer_(buffer_size_) {
    //magic_flags_ = Common::kLRS;
    TRACE("Initialized LRS");
  }

  Splitter_LRS::~Splitter_LRS() {}

  void Splitter_LRS::ProcessLostChunk(int lost_chunk_number, const boost::asio::ip::udp::endpoint &sender) {
    Splitter_DBS::ProcessLostChunk(lost_chunk_number, sender);
    std::vector<char> message = buffer_.at(lost_chunk_number % buffer_size_);
    asio::ip::udp::endpoint peer = peer_list_.at(0);
    
    system::error_code ec;
    
    // Send always to monitor peer
    team_socket_.send_to(asio::buffer(message), peer, 0, ec);
    
    if (ec) {
      ERROR("LRS - Error sending chunk: " << ec.message());
    }
    
    TRACE("Re-sending " << to_string(lost_chunk_number) << " to " << peer);
  }
  
  void Splitter_LRS::SendChunk(const std::vector<char> &message,
			       const boost::asio::ip::udp::endpoint &destination) {
    Splitter_DBS::SendChunk(message, destination);
    buffer_[chunk_number_ % buffer_size_] = message;
  }
}
